docs/source/tutorials/llama_kd_tutorial.rst

```diff
@@ -0,0 +1,229 @@
+more parameters and capacity for knowledge, however, this larger capacity is also more computationally
+expensive to deploy. Knowledge distillation can be used to compress the knowledge of a larger model into
+a smaller model. The idea is that performance of smaller models can be improved by learning from larger
+model's outputs.
+
+How does Knowledge Distillation work?
+-------------------------------------
+
+Knowledge is transferred from the teacher to student model by training it on a transfer set and where the
+student is trained to imitate the token-level probability distributions of the teacher. The diagram below
+is a simplified representation of how KD works.
+
+.. image:: /_static/img/kd-simplified.png
+
+The total loss can be configured in many ways. The default KD config in torchtune combines the cross-entropy (CE) loss with the
+forward `Kullback-Leibler (KL) divergence <https://en.wikipedia.org/wiki/Kullback%E2%80%93Leibler_divergence>`_ loss,
+which is used in standard KD approaches. Forward KL divergence aims to minimize the difference by forcing the student's
+distribution to align with all of the teacher's distributions. However, aligning the student distribution to the whole
+teacher distribution may not be effective and there are multiple papers, such as `MiniLLM <https://arxiv.org/pdf/2306.08543>`_,
+`DistiLLM <https://arxiv.org/pdf/2402.03898>`_, and `Generalized KD <https://arxiv.org/pdf/2306.13649>`_,
+that introduce new KD losses to address the limitations. For this tutorial, let's take a look at the implementation of
+the forward KL divergence loss.
+
+.. code-block:: python
+
+  import torch
+  import torch.nn.functional as F
+
+  class ForwardKLLoss(torch.nn.Module):
+    def __init__(self, ignore_index: int = -100)
+      super().__init__()
+      self.ignore_index = ignore_index
+
+    def forward(self, student_logits, teacher_logits, labels) -> torch.Tensor:
+      # Implementation from https://github.com/jongwooko/distillm
+      # Computes the softmax of the teacher logits
+      teacher_prob = F.softmax(teacher_logits, dim=-1, dtype=torch.float32)
+      # Computes the student log softmax probabilities
+      student_logprob = F.log_softmax(student_logits, dim=-1, dtype=torch.float32)
+      # Computes the forward KL divergence
+      prod_probs = teacher_prob * student_logprob
+      # Compute the sum
+      x = torch.sum(prod_probs, dim=-1).view(-1)
+      # We don't want to include the ignore labels in the average
+      mask = (labels != self.ignore_index).int()
+      # Loss is averaged over non-ignored targets
+      return -torch.sum(x * mask.view(-1), dim=0) / torch.sum(mask.view(-1), dim=0)
+
+There are some details omitted to simplify the computation, but if you'd like to know more,
+you can see the implementation in :class:`torchtune.modules.loss.ForwardKLLoss`.
+By default, the KD configs use :class:`torchtune.modules.loss.ForwardKLWithChunkedOutputLoss` to reduce memory.
```

Comment:
I think you need to add these APIs to the rst file [here](https://github.com/pytorch/torchtune/blob/4efd7fdb48677a4ac6d78f394be7b4ecfbabf7de/docs/source/api_ref_modules.rst?plain=1#L35). This way it'll render as an actual pointer to our API ref

Reply:
added. Thanks for the pointer.